### PR TITLE
Avro import

### PR DESCRIPTION
**Pull request by ericzhang6222**
* Please attach **WIP** tag when this PR is still work in progress.
* Please attach **breaking-change** tag if this PR potentially causes other components to fail or changes previous sysl executable's behaviour.

Fixes # .

Changes proposed in this pull request:
- Added arr.ai transformer for Avro to Sysl
- 

Checklist:
- [ ] Added related tests
- [ ] Made corresponding changes to the documentation

### CODE REVIEW

File: pkg/importer/avro/tests/complex_test.arrai
Comment (orlade-anz) [inline pkg/importer/avro/tests/complex_test.arrai:68]: Nit: newline

File: pkg/importer/avro/tests/complex_test.arrai
Comment (orlade-anz) [inline pkg/importer/avro/tests/complex_test.arrai:77]: Nit: if this is the name pattern we're going for, capitalise the start of each part (i.e. the T in TagMapItem). Alternatively something like String_tagMapItem_Item
 ```suggestion
    !type StringTagMapItemItem[json_map_key="key"]:
```

File: pkg/importer/avro/tests/complex_test.arrai
Comment (orlade-anz) [inline pkg/importer/avro/tests/complex_test.arrai:111]: Nit: Are these not the same thing? We shouldn't have both. If they're generated, sort the inner types by name so they end up the same thing.

File: pkg/importer/avro/tests/complex_test.arrai
Comment (orlade-anz) [inline pkg/importer/avro/tests/complex_test.arrai:103]: Nit: newline above

File: pkg/importer/avro/tests/complex_test.arrai
Comment (orlade-anz) [inline pkg/importer/avro/tests/complex_test.arrai:1]: It's hard to recognise that this file is actually a test. Import the fixtures instead, like

```
  //test.assert.equal(//os.file('complex.golden.sysl'), //{./../transformer}("complex.avsc", "Avro", "com.example"))
```

File: pkg/importer/avro/transformToSyslUnion.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformToSyslUnion.arrai:24]: Nit: space inside the `${...}` is ignore, so inline these one-liners
```suggestion
    ${types where .@item != 'null' rank (:.@) >> $'${util.indent}${.}'::}\n'
```

File: pkg/importer/avro/transformToSyslUnion.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformToSyslUnion.arrai:30]: As before

File: pkg/importer/avro/transformer.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformer.arrai:13]: Nit: The `(:s)` is binding `items.s` to a variable named `s`, so you can just use that.
```suggestion
        (:s): extraEnums(s), # primitive
```

File: pkg/importer/avro/transformer.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformer.arrai:26]: You don't use `name` here, you can get rid of it.
```suggestion
    {"type": (s: "record"), "fields": (a : fields), ...}:
```

File: pkg/importer/avro/tests/complex_test.arrai
Comment (andrewemeryanz) [reply pkg/importer/avro/tests/complex_test.arrai:17]: It's correct in this instance. In Avro, the order of the union is important. `null` is usually the first item in the union so we only append an attribute to capture the order if this isn't the case.

File: pkg/importer/avro/tests/complex_test.arrai
Comment (andrewemeryanz) [inline pkg/importer/avro/tests/complex_test.arrai:8]: I don't think these section headers are necessary.

File: pkg/importer/avro/tests/complex_test.arrai
Comment (andrewemeryanz) [reply pkg/importer/avro/tests/complex_test.arrai:58]: Avro types can have no default value or a variety of complex default values depending on the type. We're not inspecting the default values, just noting their value in string form.

**Comment by ericzhang6222** [comment]
Can't merge until #976 is merged.

File: pkg/importer/avro/tests/complex_test.arrai
Comment (andrewemeryanz) [reply pkg/importer/avro/tests/complex_test.arrai:111]: Good question.

The order of the types within the Sysl union is being used to represent the order of the types within Avro union. The `union` attribute is only added if the order can't be derived by the type (and nullable value) alone. 

I'm tempted to go with this suggestion. It will likely add additional `union` attributes in the general case but I think it's a cleaner representation. @ericzhang6222 I'll update specification.

File: pkg/importer/avro/util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/util.arrai:109]:
```suggestion
                'array': 'sequence of ' + cond type('items') {{'name': (:s), ...}: s, (:s): s},
                'map': 'set of ' + cond type('items') {{'name': (:s), ...}: 'String' + s + 'Item', (:s): s},
```

File: pkg/importer/avro/tests/complex_test.arrai
Comment (andrewemeryanz) [reply pkg/importer/avro/tests/complex_test.arrai:111]: @ericzhang6222 Specification updated.

File: pkg/importer/avro/transformToSyslUnion.arrai
Comment (ericzhang6222) [reply pkg/importer/avro/transformToSyslUnion.arrai]: Both `fullTypeNames` and `types` are used in following code.

File: pkg/importer/avro/tests/complex_test.arrai
Comment (ericzhang6222) [reply pkg/importer/avro/tests/complex_test.arrai]: This comes from https://github.com/anzx/acceleration/issues/328
```
sysl import required flags:

-i, --input=<SCHEMA> The schema to import
-a, --app-name=<APP-NAME> The name of the Sysl app.
sysl import optional flags:

-p, --package=<PACKAGE> The name of the Sysl package.
-o, --output=<OUT.SYSL> The output filename.
```

File: pkg/importer/avro/tests/complex_test.arrai
Comment (ericzhang6222) [reply pkg/importer/avro/tests/complex_test.arrai]: @andrewemeryanz do you want updating to `@description` syntax?

File: pkg/importer/avro/transformToSyslType.arrai
Comment (ericzhang6222) [reply pkg/importer/avro/transformToSyslType.arrai]: If types is `['null', 'string']`, it outputs code `attr1 <: string?`. Or outputs code `attr <: string? [union=['string', 'null']]`.
https://confluence.service.anz/display/ABT/Avro+representation#Avrorepresentation-unions

File: pkg/importer/avro.go
Comment (orlade-anz) [inline pkg/importer/avro.go:32]: Nit: nested Sprintf is hard to follow. Keep it simple.
```suggestion
	call := fmt.Sprintf("(`%s`, `%s`, `%s`)", avroSpec, i.appName, i.pkg)
	val, err := syntax.EvaluateExpr("", fmt.Sprintf("%s%s", avro.AvroTransformerScript, call))
```

File: pkg/importer/avro.go
Comment (orlade-anz) [inline pkg/importer/avro.go:12]: Nit: full stops at the ends of comments (and below).

File: pkg/importer/avro.go
Comment (orlade-anz) [inline pkg/importer/avro.go:45]: Nit: Comments should be descriptive (describing what it does), not prescriptive (telling it what to do, like a command). So "Sets" instead of "Set".

```suggestion
// Sets the package attribute of the imported app.
```

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:141]: Nit: no spaces before `:bindings`.
```suggestion
    :prettyAnnotations,
    :transformType,
    :getAnnotationS,
    :getTypeName,
    :getDefaultVal,
    :combineTypes,
    :correctName,
```

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:134]: Nit: No need to assign a name to this tuple, just return it.
```suggestion
(
```

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:121]: Nit: Avoid verbs in function names for simple transformations. This would read better as just `defaultVal`, since that's what it returns. What it does (i.e. `get`) is not interesting in the context of the calling code.

Also avoid using `\.` in most cases. It works, but you're passing up an opportunity to document the data you're working with. What kind of input is this function expecting? kinds of thing have a "defaultVal"? I have to be familiar with Avro to know that, and I shouldn't need to know that to understand your code.
```suggestion
let defaultVal = \.
```

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:56]: Nit: Group this function with `transformPrimitiveType`, since it appears to be the only user.

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:28]: Nit: consistent quotes
```suggestion
        'bytes': 'bytes',
```

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:14]: Nit: simpler name, same information
```suggestion
let prettyAnnotations = \annotations
```

File: pkg/importer/avro/avro_util.arrai
Comment (orlade-anz) [inline pkg/importer/avro/avro_util.arrai:13]: What does this mean? Provides no more information than the function name

**Review by orlade-anz — CHANGES_REQUESTED** [review]

File: pkg/importer/avro/to_sysl_enum.arrai
Comment (orlade-anz) [inline pkg/importer/avro/to_sysl_enum.arrai:1]:
```suggestion
# Transforms an Avro enum to a Sysl enum.
```

File: pkg/importer/avro/to_sysl_type.arrai
Comment (orlade-anz) [inline pkg/importer/avro/to_sysl_type.arrai:1]:
```suggestion
# Transforms an Avro record to a Sysl type.
```

File: pkg/importer/avro/to_sysl_type.arrai
Comment (orlade-anz) [inline pkg/importer/avro/to_sysl_type.arrai:55]: Ditto, indenting.

File: pkg/importer/avro/to_sysl_type.arrai
Comment (orlade-anz) [inline pkg/importer/avro/to_sysl_type.arrai:97]: Nit: leave some whitespace before returning.
```suggestion
}}`}}`;

transformer
```

File: pkg/importer/avro/transformer.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformer.arrai:3]: I'm confused as to why these are "extra". What does "extra" mean in this context?

File: pkg/importer/avro/to_sysl_union.arrai
Comment (orlade-anz) [inline pkg/importer/avro/to_sysl_union.arrai:21]:
```suggestion
;

transformer
```

File: pkg/importer/avro/transformer.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformer.arrai:105]:
```suggestion

avroTransformer
```

File: pkg/importer/avro/transformer.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformer.arrai:95]:
```suggestion
${appName}${util.prettyAnnotations(['avro_spec="1.0"', $`${cond packageName {'':'', _:$`package="${packageName}"`}}`])}:
```

File: pkg/importer/avro/transformer.arrai
Comment (orlade-anz) [inline pkg/importer/avro/transformer.arrai:103]: Nit: newlines should be consistent.
```suggestion
    ${extraTypes(schema) => //{./to_sysl_type}(.) orderby . ::\n}

    # Aliases
    ${extraAliases(schema) => //{./to_sysl_alias}(.) orderby . ::\n}

    # Unions
    ${extraUnions(schema) => //{./to_sysl_union}(.) orderby . ::\n}

    # Enums
    ${extraEnums(schema) => //{./to_sysl_enum}(.) orderby . ::\n}
```

File: pkg/importer/formats.go
Comment (orlade-anz) [inline pkg/importer/formats.go:36]:
```suggestion
var Avro = Format{
```

**Review by orlade-anz — CHANGES_REQUESTED** [review]
More nits, mostly around whitespace :)

File: pkg/importer/avro.go
Comment (anzdaddy) [inline pkg/importer/avro.go:30]: Building parameters as strings is fragile. We should define a call interface that takes regular Go types and converts them to appropriate arr.ai types and passes them as parameters. It should end up something like:

```
rel.Call(syntax.EvaluateExpr(avro.AvroTransformerScript), avroSpec, i.appName, i.pkg)
```

File: pkg/importer/avro/tests/complex_test.arrai
Comment (anzdaddy) [inline pkg/importer/avro/tests/complex_test.arrai]: Why does this type exist? It isn't used anywhere.

**Comment by orlade-anz** [comment]
Merged internally to be pushed out.

**Pull request closed without merging**
